client/src/main/java/io/confluent/kafka/schemaregistry/client/CachedSchemaRegistryClient.java

```diff
@@ -247,21 +242,19 @@ public synchronized int getVersion(String subject, Schema schema)
   @Override
   public synchronized int getId(String subject, Schema schema)
       throws IOException, RestClientException {
-    Map<Schema, Integer> schemaIdMap;
+    IdentityLRUCache<Schema, Integer> schemaIdMap;
     if (schemaCache.containsKey(subject)) {
       schemaIdMap = schemaCache.get(subject);
     } else {
-      schemaIdMap = new IdentityHashMap();
+      schemaIdMap = new IdentityLRUCache<>(identityMapCapacity);
       schemaCache.put(subject, schemaIdMap);
     }
 
-    if (schemaIdMap.containsKey(schema)) {
-      return schemaIdMap.get(schema);
+    Integer id = schemaIdMap.get(schema);
+    if (id != null) {
+      return id;
     } else {
-      if (schemaIdMap.size() >= identityMapCapacity) {
-        throw new IllegalStateException("Too many schema objects created for " + subject + "!");
-      }
-      int id = getIdFromRegistry(subject, schema);
+      id = getIdFromRegistry(subject, schema);
       schemaIdMap.put(schema, id);
       idCache.get(null).put(id, schema);
```

Comment:
Our current upper bound on the number of schemas also limits the number of entries put into the idCache for the null subject here. Now that map could potentially grow indefinitely?